### DISCUSSION TITLE
How do you deallocate used memory?

### DISCUSSION THREAD

**Discussion by NullSoldier** [samizdatco/skia-canvas#70] [Q&A]
Creating and rendering an image seems to allocate memory, but there's no obvious way to deallocate it. Clearing references to the canvas do nothing. Setting width / height to 0 do nothing. The memory is still allocated.

```javascript
const { Canvas } = require('skia-canvas')

async function printMem(name) {
    await new Promise((r) => setTimeout(r, 1000))
    global.gc()

    const mem = process.memoryUsage()
    const rss = (mem.rss/1024/1024).toFixed(0).padStart(3)
    const heap = (mem.heapUsed/1024/1024).toFixed(0).padStart(3)
    const external = (mem.external/1024/1024).toFixed(0).padStart(3)
    console.log(`${rss} MiB / ${heap} MiB / ${external} MiB / ${name}`)
}

async function run() {
    const width = 5000
    const height = 5000

    console.log('    rss /    heap /     ext / name')
    await printMem('start')

    let canvas = new Canvas(width, height)
    await printMem('canvas created')

    let context = canvas.getContext('2d')
    await printMem('context created')

    context.fillStyle = 'red'
    context.fillRect(0, 0, width, height)
    canvas.toDataURL()
    await printMem('RENDERED')

    canvas.width = 0
    canvas.height = 0
    context = null
    canvas = null
    await printMem('closed')
}

run()
//     rss /    heap /     ext / name
//  31 MiB /   3 MiB /   1 MiB / start
//  31 MiB /   3 MiB /   1 MiB / canvas created
//  31 MiB /   3 MiB /   1 MiB / context created
// 128 MiB /   3 MiB /   1 MiB / RENDERED
// 127 MiB /   3 MiB /   1 MiB / closed
```

**Comment by takase1121** [comment]
try waiting for the canvas to go out of scope and GCed.

**Comment by samizdatco** [comment]
In addition to @takase1121's point about how scope & GC interact, you should also keep in mind that the number that will fluctuate due to the objects *your* code creates are in the heap column. The [RSS](https://en.wikipedia.org/wiki/Resident_set_size) value is more of a high-water-mark than something that grows & shrinks  since it is also measuring all the shared library code that gets mapped in, etc.

If you print out the totals in kilobytes instead of megs, you'll see that there actually is some reclaiming of heap space going on:
```
    rss /     heap /     ext / name
 31 MiB / 3560 KiB / 850 KiB / start
 32 MiB / 3558 KiB / 822 KiB / canvas created
 32 MiB / 3559 KiB / 822 KiB / context created
127 MiB / 3736 KiB / 930 KiB / RENDERED
126 MiB / 3562 KiB / 822 KiB / closed
```

**Comment by luxocrates** [comment]
Let me tag newly-opened issue https://github.com/samizdatco/skia-canvas/issues/284 here. Sustained use appears to show that some of skia-canvas's memory allocations are out of the garbage collector's reach, including `Canvas` creation.
